modules/cudacodec/include/opencv2/cudacodec.hpp

```diff
@@ -66,181 +66,288 @@ using namespace cuda;  // Stream
+};
+CV_EXPORTS bool operator==(const EncoderParams& lhs, const EncoderParams& rhs);
 
-    /** @brief Callback function to signal the start of bitstream that is to be encoded.
+/** @brief Interface for encoder callbacks.
 
-    Callback must allocate buffer for CUDA encoder and return pointer to it and it's size.
-     */
-    virtual uchar* acquireBitStream(int* bufferSize) = 0;
+User can implement own multiplexing by implementing this interface.
+*/
+class CV_EXPORTS_W EncoderCallback {
+public:
+    /** @brief Callback function to signal that the encoded bitstream for one or more frames is ready.
 
-    /** @brief Callback function to signal that the encoded bitstream is ready to be written to file.
+    @param vPacket The raw bitstream for one or more frames.
     */
-    virtual void releaseBitStream(unsigned char* data, int size) = 0;
-
-    /** @brief Callback function to signal that the encoding operation on the frame has started.
-
-    @param frameNumber
-    @param picType Specify frame type (I-Frame, P-Frame or B-Frame).
-     */
-    CV_WRAP virtual void onBeginFrame(int frameNumber, EncoderCallBack::PicType picType) = 0;
+    virtual void onEncoded(std::vector<std::vector<uint8_t>> vPacket) = 0;
 
-    /** @brief Callback function signals that the encoding operation on the frame has finished.
+    /** @brief Callback function to that the encoding has finished.
+    * */
+    virtual void onEncodingFinished() = 0;
 
-    @param frameNumber
-    @param picType Specify frame type (I-Frame, P-Frame or B-Frame).
-     */
-    CV_WRAP virtual void onEndFrame(int frameNumber, EncoderCallBack::PicType picType) = 0;
+    virtual ~EncoderCallback() {}
 };
 
 /** @brief Video writer interface.
-
-The implementation uses H264 video codec.
-
-@note Currently only Windows platform is supported.
-
 @note
    -   An example on how to use the videoWriter class can be found at
         opencv_source_code/samples/gpu/video_writer.cpp
- */
+*/
 class CV_EXPORTS_W VideoWriter
 {
 public:
     virtual ~VideoWriter() {}
 
     /** @brief Writes the next video frame.
 
-    @param frame The written frame.
-    @param lastFrame Indicates that it is end of stream. The parameter can be ignored.
+    @param frame The framet to be written.
 
-    The method write the specified image to video file. The image must have the same size and the same
+    The method encodes the specified image to a video stream. The image must have the same size and the same
     surface format as has been specified when opening the video writer.
-     */
-    CV_WRAP virtual void write(InputArray frame, bool lastFrame = false) = 0;
+    */
+    CV_WRAP virtual void write(InputArray frame) = 0;
 
+    /** @brief Retrieve the encoding parameters.
+    */
     CV_WRAP virtual EncoderParams getEncoderParams() const = 0;
+
+    /** @brief Waits until the encoding process has finished before calling EncoderCallback::onEncodingFinished().
+    */
+    CV_WRAP virtual void close() = 0;
```

Comment:
I propose to name the method `release()` to be consistent with `cv::VideoWriter` API.